### PR TITLE
Create reports_definition.py for launching reports from pyTenable

### PR DESCRIPTION
**Pull request by samwmarsh**
Tenable SC launch reports using ID in pytenable

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* Python Version(s) Tested:
* Tenable.sc version (if necessary):

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by aseemsavio — COMMENTED** [review]
Could you provide information on how this was tested? You can refer to existing tests in this directory `tests/sc/`.

**Comment by aseemsavio** [comment]
Hey @samwmarsh no problem at all! We include tests that confirms the behaviour of a new feature we're adding with the respective pull request. Give me some time, I'll let you know the changes required.

File: tenable/sc/__init__.py [hunk 1]
```diff
@@ -430,6 +432,14 @@ def recast_risks(self):
         '''
         return RecastRiskAPI(self)
 
+    @property
+    def reports_definition(self):
```
Comment (SteveMcGrath) [inline tenable/sc/__init__.py:436]: Should be using a more correct name that conforms to the rest of the package

```suggestion
    def report_definitions(self):
```

File: tenable/sc/reports_definition.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+Methods available on ``sc.reports_definition``:
+
+.. rst-class:: hide-signature
+.. autoclass:: ReportsDefinitionAPI
+    :members:
+'''
+from .base import SCEndpoint
+
+class ReportsDefinitionAPI(SCEndpoint):
```
Comment (SteveMcGrath) [inline tenable/sc/reports_definition.py:16]: Should be using a more correct name that conforms to the rest of the package

```suggestion
class ReportDefinitionsAPI(SCEndpoint):
```

File: tenable/sc/__init__.py [hunk 3]
```diff
@@ -65,6 +66,7 @@
 from .policies import ScanPolicyAPI
 from .queries import QueryAPI
 from .recast_risks import RecastRiskAPI
+from .reports_definition import ReportsDefinitionAPI
```
Comment (SteveMcGrath) [inline tenable/sc/__init__.py:69]: Should be using a more correct name that conforms to the rest of the package

```suggestion
from .report_definitions import ReportDefinitionsAPI
```

File: tenable/sc/reports_definition.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    def launch(self, id):
+        '''
+        Launches a Report definition.
+        :sc-api:`reports-definition: launch <ReportsDefinition.html#ReportsDefinitionRESTReference-/reportDefinition/{id}/launch>`
```
Comment (SteveMcGrath) [inline tenable/sc/reports_definition.py:20]:
```suggestion
        :sc-api:`report-definitions: launch <ReportsDefinition.html#ReportsDefinitionRESTReference-/reportDefinition/{id}/launch>`
```

File: tenable/sc/reports_definition.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        Args:
+            id (int): The report definition identifier to launch.
+        Returns:
+            :obj:`dict`:
+                A report ID resource for the newly launched report definition.
+        Examples:
+            >>> running = sc.reports_definition.launch(1)
```
Comment (SteveMcGrath) [inline tenable/sc/reports_definition.py:27]:
```suggestion
            >>> running = sc.report_definitions.launch(1)
```

File: tenable/sc/__init__.py [hunk 6]
```diff
@@ -430,6 +432,14 @@ def recast_risks(self):
         '''
         return RecastRiskAPI(self)
 
+    @property
+    def reports_definition(self):
+        '''
+        The interface object for the
+        :doc:`Tenable.sc ReportDefinition APIs <reports_definition>`.
+        '''
+        return ReportsDefinitionAPI(self)
```
Comment (SteveMcGrath) [inline tenable/sc/__init__.py:441]:
```suggestion
        return ReportDefinitionsAPI(self)
```

**Review by SteveMcGrath — CHANGES_REQUESTED** [review]

**Comment by samwmarsh** [comment]
@aseemsavio all changes merged thanks

**Comment by samwmarsh** [comment]
@SteveMcGrath that's all sorted, anything blocking this merge now?

**Comment by SteveMcGrath** [comment]
I'd say there are still some issues just looking at the test suite.

https://github.com/tenable/pyTenable/actions/runs/3881092089/jobs/6619727711#step:5:258

**Comment by samwmarsh** [comment]
@SteveMcGrath sorry for the delay on this one, I think I've resolved the test suite issues - seemed to be an issue with typo between file name and imports.

**Comment by samwmarsh** [comment]
@SteveMcGrath anything this is waiting on?

**Review by SteveMcGrath — COMMENTED** [review]
Aside from missing unit tests, looks good.

**Comment by SteveMcGrath** [comment]
Apologize for the delay, this one got pushed to the bottom of the stack.  if you're willing to write the unit tests, I have no issues getting this approved.

**Comment by samwmarsh** [comment]
Not a python dev by any means but hows that? all checks seem happy.

**Comment by SteveMcGrath** [comment]
The two tests you wrote look great, just need a test with a success as well.  Not anything crazy, we just need to assert that were getting a valid response.  maybe check that the ID exists in the response and it should be ready to merge.

**Comment by samwmarsh** [comment]
Something like that?

**Review by SteveMcGrath — APPROVED** [review]

**Comment by samwmarsh** [comment]
```FAILED tests/sc/test_report_definition.py::test_report_definition_launch_success - requests.exceptions.ConnectionError: HTTPSConnectionPool(host='securitycenter.home.cugnet.net', port=443): Max retries exceeded with url: /rest/reportDefinition/1/launch (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f63a12b2990>: Failed to establish a new connection: [Errno -2] Name or service not known'))```

I copied this from a different test, is this ok or is there a host value this should be to pass?

File: tests/sc/test_report_definition.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+def test_report_definition_id_item_typeerror(security_center):
+    '''
+    validate id is int not list
+    '''
+    
+    with pytest.raises(TypeError):
+        security_center.report_definition.launch(id=['one'])
+
+def test_report_definition_launch_success(security_center):
```
Comment (SteveMcGrath) [inline tests/sc/test_report_definition.py:22]: Solves https://github.com/tenable/pyTenable/actions/runs/11962140227/job/33349851654?pr=656#step:7:808

```suggestion

@pytest.mark.vcr()
def test_report_definition_launch_success(security_center):
```

File: tests/sc/cassettes/test_report_definition_launch_success.yaml [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+      keep-alive: ['timeout=15, max=100']
+      pragma: [no-cache]
+      securitycenter: [5.8.0]
+      server: [Apache]
+      strict-transport-security: [max-age=31536000; includeSubDomains]
+      x-content-type-options: [nosniff]
+      x-frame-options: [DENY]
+      x-xss-protection: [1; mode=block]
+    status: {code: 200, message: OK}
+version: 1
```
Comment (SteveMcGrath) [inline tests/sc/cassettes/test_report_definition_launch_success.yaml:38]:
```suggestion
interactions:
- request:
    body: null
    headers:
      Accept: ['*/*']
      Accept-Encoding: ['gzip, deflate']
      Connection: [keep-alive]
      Cookie: [TNS_SESSIONID=SESSIONID]
      User-Agent: [pyTenable/1.5.3 (pyTenable/1.5.3; Python/3.9.13)]
      X-SecurityCenter: ['0000000000']
    method: POST
    uri: https://securitycenter.home.cugnet.net/rest/reportDefinition/1/launch
  response:
    body: {string: !!python/unicode '{"type" : "regular","response" : {"reportDefinitionID" : "1","reportResult" : {"creatorID" : "2","ownerID" : "2","ownerGID" : "1","reportDefinitionID" : "7","jobID" : "113868","name" : "CS-4700 Iterator","description" : "","type" : "pdf","status" : "Queued","errorDetails" : "","totalSteps" : -1,"completedSteps" : 0,"startTime" : -1,"finishTime" : 0,"pubSites" : [],"id" : "1"}}, "error_code" : 0, "error_msg" : "", "warnings" : [], "timestamp" : 1434567869}'}
    headers:
      cache-control: ['no-store, no-cache, must-revalidate']
      connection: [Keep-Alive]
      content-length: ['983']
      content-type: [application/json]
      date: ['Mon, 08 Apr 2019 13:56:30 GMT']
      expires: ['Thu, 19 Nov 1981 08:52:00 GMT']
      keep-alive: ['timeout=15, max=100']
      pragma: [no-cache]
      securitycenter: [5.8.0]
      server: [Apache]
      strict-transport-security: [max-age=31536000; includeSubDomains]
      x-content-type-options: [nosniff]
      x-frame-options: [DENY]
      x-xss-protection: [1; mode=block]
    status: {code: 200, message: OK}
version: 1
```

File: tests/sc/test_report_definition.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        security_center.report_definition.launch(id=['one'])
+
+@pytest.mark.vcr()
+def test_report_definition_launch_success(security_center):
+    '''
+    validate launch is successful and returns running report ID
+    '''
+    report = security_center.report_definition.launch(id=1)
+    assert isinstance(report, dict)
+    check(report['reportResult'], 'id', int)
```
Comment (SteveMcGrath) [inline tests/sc/test_report_definition.py:29]:
```suggestion
    check(report['reportResult'], 'id', str)
```

**Comment by samwmarsh** [comment]
Thanks @SteveMcGrath, apologies I've not got access to .sc from my public github location so it was guesswork based off of the api docs!

**Comment by SteveMcGrath** [comment]
Not a problem, it appeared that the JSON you posted in the cassette wasn't valid, so i tweaked it.

**Pull request merged**